File: 351004/Brazhalovich/Lab_1/publisher/src/main/java/org/example/newsapi/controller/AuthController.java
package org.example.newsapi.controller;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.example.newsapi.config.JwtTokenProvider;
import org.example.newsapi.dto.request.LoginRequest;
import org.example.newsapi.dto.request.UserRegistrationRequest;
import org.example.newsapi.dto.response.UserResponseTo;
import org.example.newsapi.entity.User;
import org.example.newsapi.exception.AlreadyExistsException;
import org.example.newsapi.mapper.UserMapper;
import org.example.newsapi.repository.UserRepository;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/api/v2.0")
@RequiredArgsConstructor
public class AuthController {

    private final AuthenticationManager authenticationManager;
    private final JwtTokenProvider tokenProvider;
    private final UserRepository userRepository;
    private final UserMapper userMapper;
    private final PasswordEncoder passwordEncoder;

    @PostMapping("/login")
    public ResponseEntity<?> login(@RequestBody @Valid LoginRequest request) {
        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(request.getLogin(), request.getPassword())
        );
        String role = authentication.getAuthorities().iterator().next().getAuthority().replace("ROLE_", "");
        String token = tokenProvider.generateToken(request.getLogin(), role);
        return ResponseEntity.ok(Map.of("access_token", token));
    }

    @PostMapping("/users")
    @ResponseStatus(HttpStatus.CREATED)
    public UserResponseTo register(@RequestBody @Valid UserRegistrationRequest request) {
        if (userRepository.existsByLogin(request.getLogin())) {
            throw new AlreadyExistsException("Login already exists");
        }
        User user = User.builder()
                .login(request.getLogin())
                .password(passwordEncoder.encode(request.getPassword()))
                .firstname(request.getFirstname())
                .lastname(request.getLastname())
                .role(request.getRole() != null ? request.getRole() : "CUSTOMER")
                .build();
        user = userRepository.save(user);
        return userMapper.toDto(user);
    }
}

File: 351003/brezgunov/distcomp/src/main/kotlin/com/distcomp/controller/AuthController.kt
package com.distcomp.controller

import com.distcomp.dto.auth.LoginRequest
import com.distcomp.dto.auth.LoginResponse
import com.distcomp.repository.UserRepository
import com.distcomp.security.JwtTokenProvider
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v2.0")
class AuthController(
    private val authenticationManager: AuthenticationManager,
    private val jwtTokenProvider: JwtTokenProvider,
    private val userRepository: UserRepository
) {
    @PostMapping("/login")
    fun login(@RequestBody request: LoginRequest): LoginResponse {
        authenticationManager.authenticate(
            UsernamePasswordAuthenticationToken(request.login, request.password)
        )
        val user = userRepository.findByLogin(request.login)!!
        val token = jwtTokenProvider.generateToken(user.login, user.role.name)
        return LoginResponse(access_token = token)
    }
}